File: app/src/main/java/com/itba/runningMate/map/Map.java
package com.itba.runningMate.map;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;

import androidx.annotation.NonNull;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMapOptions;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.PolylineOptions;
import com.itba.runningMate.domain.Route;

import java.util.List;

import static com.itba.runningMate.Constants.DEFAULT_LATITUDE;
import static com.itba.runningMate.Constants.DEFAULT_LONGITUDE;
import static com.itba.runningMate.Constants.DEFAULT_ZOOM;
import static com.itba.runningMate.Constants.MY_LOCATION_ZOOM;

public class Map extends MapView {

    private static final int PADDING = 20; // padding de los puntos en el mapa

    private GoogleMap googleMap;

    public Map(Context context) {
        super(context);
    }

    public Map(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
    }

    public Map(Context context, AttributeSet attributeSet, int i) {
        super(context, attributeSet, i);
    }

    public Map(Context context, GoogleMapOptions googleMapOptions) {
        super(context, googleMapOptions);
    }

    public void bind(@NonNull GoogleMap googleMap) {
        this.googleMap = googleMap;
    }

    public void showLocation(double latitude, double longitude, float zoom) {
        if (googleMap != null) {
            googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(latitude, longitude), zoom));
        }
    }

    public void showDefaultLocation() {
        showLocation(DEFAULT_LATITUDE, DEFAULT_LONGITUDE, DEFAULT_ZOOM);
    }

    public void showLocation(double latitude, double longitude) {
        showLocation(latitude, longitude, MY_LOCATION_ZOOM);
    }

    public void setOnCameraMoveStartedListener(GoogleMap.OnCameraMoveStartedListener mapCameraListener) {
        if (googleMap != null) {
            googleMap.setOnCameraMoveStartedListener(mapCameraListener);
        }
    }

    public void setOnMyLocationButtonClickListener(GoogleMap.OnMyLocationButtonClickListener mapMyLocationButtonListener) {
        if (googleMap != null) {
            googleMap.setOnMyLocationButtonClickListener(mapMyLocationButtonListener);
        }
    }

    public void setCompassEnabled(boolean isCompassEnabled) {
        if (googleMap != null) {
            googleMap.getUiSettings().setCompassEnabled(true);
        }
    }

    @SuppressLint("MissingPermission")
    public void enableMyLocation() {
        if (googleMap != null) {
            googleMap.setMyLocationEnabled(true);
            googleMap.getUiSettings().setMyLocationButtonEnabled(true);
        }
    }

    @SuppressLint("MissingPermission")
    public void disableMyLocation() {
        if (googleMap != null) {
            googleMap.setMyLocationEnabled(false);
            googleMap.getUiSettings().setMyLocationButtonEnabled(false);
        }
    }

    public void showRoute(Route route) {
        if (route.isEmpty() || googleMap == null) {
            return;
        }
        for (List<LatLng> lap : route.getLocations()) {
            googleMap.addPolyline(new PolylineOptions()
                    .color(Color.BLUE)
                    .width(8f)
                    .addAll(lap));
        }
    }

    public void centerMapOn(Route route) {
        if (route == null || route.isEmpty()) {
            return;
        }

        LatLngBounds.Builder boundsBuilder = LatLngBounds.builder();
        for (List<LatLng> lap : route.getLocations()) {
            for (LatLng point : lap) {
                boundsBuilder.include(point);
            }
        }
        googleMap.moveCamera(CameraUpdateFactory.newLatLngBounds(boundsBuilder.build(), PADDING));
    }
}

File: app/src/main/java/com/itba/runningMate/map/Map.kt
package com.itba.runningMate.map

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.OnCameraMoveStartedListener
import com.google.android.gms.maps.GoogleMap.OnMyLocationButtonClickListener
import com.google.android.gms.maps.GoogleMapOptions
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.model.*
import com.itba.runningMate.domain.Route
import com.itba.runningMate.utils.Constants.DEFAULT_LATITUDE
import com.itba.runningMate.utils.Constants.DEFAULT_LONGITUDE
import com.itba.runningMate.utils.Constants.DEFAULT_ZOOM
import com.itba.runningMate.utils.Constants.MY_LOCATION_ZOOM

open class Map : MapView {
    private var googleMap: GoogleMap? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attributeSet: AttributeSet) : super(context, attributeSet)

    constructor(context: Context, attributeSet: AttributeSet, i: Int) : super(
        context,
        attributeSet,
        i
    )

    constructor(context: Context, googleMapOptions: GoogleMapOptions) : super(
        context,
        googleMapOptions
    )

    fun bind(googleMap: GoogleMap) {
        this.googleMap = googleMap
    }

    @JvmOverloads
    fun showLocation(
        latitude: Double,
        longitude: Double,
        zoom: Float = MY_LOCATION_ZOOM.toFloat()
    ) {
        if (googleMap != null) {
            googleMap!!.animateCamera(
                CameraUpdateFactory.newLatLngZoom(
                    LatLng(latitude, longitude),
                    zoom
                )
            )
        }
    }

    fun showDefaultLocation() {
        showLocation(DEFAULT_LATITUDE, DEFAULT_LONGITUDE, DEFAULT_ZOOM.toFloat())
    }

    fun setOnCameraMoveStartedListener(mapCameraListener: OnCameraMoveStartedListener?) {
        if (googleMap != null) {
            googleMap!!.setOnCameraMoveStartedListener(mapCameraListener)
        }
    }

    fun setOnMyLocationButtonClickListener(mapMyLocationButtonListener: OnMyLocationButtonClickListener?) {
        if (googleMap != null) {
            googleMap!!.setOnMyLocationButtonClickListener(mapMyLocationButtonListener)
        }
    }

    fun setCompassEnabled(isCompassEnabled: Boolean) {
        if (googleMap != null) {
            googleMap!!.uiSettings.isCompassEnabled = isCompassEnabled
        }
    }

    @SuppressLint("MissingPermission")
    fun enableMyLocation() {
        if (googleMap != null) {
            googleMap!!.isMyLocationEnabled = true
            googleMap!!.uiSettings.isMyLocationButtonEnabled = true
        }
    }

    @SuppressLint("MissingPermission")
    fun disableMyLocation() {
        if (googleMap != null) {
            googleMap!!.isMyLocationEnabled = false
            googleMap!!.uiSettings.isMyLocationButtonEnabled = false
        }
    }

    fun showRoute(route: Route) {
        if (route.isEmpty() || googleMap == null) {
            return
        }
        for (lap in route.getLocations()) {
            googleMap!!.addPolyline(
                PolylineOptions()
                    .color(Color.BLUE)
                    .width(8f)
                    .addAll(lap)
            )
        }
    }

    fun showRouteWithMarker(route: Route) {
        showRoute(route)
        val start = LatLng(route.firstLatitude(), route.firstLongitude())
        val end = LatLng(route.lastLatitude(), route.lastLongitude())
        googleMap!!.addMarker(
            MarkerOptions()
                .position(start)
                .icon(
                    BitmapDescriptorFactory
                        .defaultMarker(BitmapDescriptorFactory.HUE_GREEN)
                )
        )
        googleMap!!.addMarker(
            MarkerOptions()
                .position(end)
                .icon(
                    BitmapDescriptorFactory
                        .defaultMarker(BitmapDescriptorFactory.HUE_RED)
                )
        )
    }

    fun centerMapOn(route: Route?) {
        if (route == null || route.isEmpty()) {
            return
        }
        val boundsBuilder = LatLngBounds.builder()
        for (lap in route.getLocations()) {
            for (point in lap) {
                boundsBuilder.include(point)
            }
        }
        googleMap!!.moveCamera(CameraUpdateFactory.newLatLngBounds(boundsBuilder.build(), PADDING))
    }

    companion object {
        private const val PADDING = 100
    }
}